Update CUDA_UPGRADE_GUIDE.MD (#998)
pytorch/builder: CUDA_UPGRADE_GUIDE.MD

## 1. Maintain Progress and Updates
Make an issue to track the progress, for example [#56721: Support 11.3](https://github.com/pytorch/pytorch/issues/56721). This is especially important as many PyTorch external users are interested in CUDA upgrades.

## 2. Modify scripts to install the new CUDA for Conda Docker Linux containers.

> There are three types of Docker containers we maintain in order to build Linux binaries: `conda`, `libtorch`, and `manywheel`. They all require installing CUDA and then updating code references in respective build scripts/Dockerfiles.  This step is about conda

Reviewer: ```suggestion
There are three types of Docker containers we maintain in order to build Linux binaries: `conda`, `libtorch`, and `manywheel`. They all require installing CUDA and then updating code references in respective build scripts/Dockerfiles.  This step is about conda.
```

1. Modifying [`install_cuda.sh`](common/install_cuda.sh) and build Conda Docker file:
    1. Follow this [PR 992](https://github.com/pytorch/builder/pull/992) for all steps in this section
    2. Find the CUDA install link [here](https://developer.nvidia.com/cuda-downloads?target_os=Linux&target_arch=x86_64&=Debian&target_version=10&target_type=runfile_local)
    3. Get the cudnn link from NVIDIA on the PyTorch Slack
    4. Run the `install_116` chunk of code on your devbox to make sure it works.
    5. Check [this link](https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/) to see if you need to add/remove any architectures to the nvprune list.
    6. Go into your cuda-11.6 folder and make sure what you're pruning actually exists. Update versions as needed, especially the visual tools like `nsight-systems`.
    7. Add setup for our Docker `conda` scripts/Dockerfiles

>     8. To test that your code works, from the root builder repo, run something similar to `export CUDA_VERSION=11.3 && ./conda/build_docker.sh` for the `conda` images. 

Reviewer: ```suggestion
    8. To test that your code works, from the root builder repo, run something similar to `export CUDA_VERSION=11.6 && ./conda/build_docker.sh` for the `conda` images. 
```
Maybe switch out all the 11.blah to 11.x for better upkeep though CUDA 12 might be happening soon so maybe not

Author: lets keep 11.6 references for now. I want to update the readme to all latest PRs

    9. Validate conda-builder docker hub [cuda11.6](https://hub.docker.com/r/pytorch/conda-builder/tags?page=1&name=cuda11.6) to see that images have been built and correctly tagged. These images are used in the next step to build Magma for linux.

## 3. Update Magma for Linux

> 1. Build Magma for linux. Our linux CUDA jobs use conda, so we need to build magma-cuda116 and push it to anaconda:

Reviewer: ```suggestion
1. Build Magma for Linux. Our Linux CUDA jobs use conda, so we need to build magma-cuda116 and push it to anaconda:
```
nits

    1. Follow this [PR 997](https://github.com/pytorch/builder/pull/997) for all steps in this section
    2. Currently, this is mainly copy-paste in [`magma/Makefile`](magma/Makefile) if there are no major code API changes/deprecations to the CUDA version. Previously, we've needed to add patches to MAGMA, so this may be something to check with NVIDIA about.
    3. To push the package, please update build-magma-linux workflow [PR 897](https://github.com/pytorch/builder/pull/897).

>     4. NOTE: This step relies on the conda-builder image (changes to `.github/workflows/build-conda-images.yml`), so make sure you have pushed the new conda-builder prior. Validate this step by logging into anaconda.org and seeing your package deployed for example [here](https://anaconda.org/pytorch/magma-cuda115)

Reviewer: ```suggestion
    4. NOTE: This step relies on the conda-builder image (changes to `.github/workflows/build-conda-images.yml`), so make sure you have pushed the new conda-builder prior. Validate this step by logging into anaconda.org and seeing your package deployed for example [here](https://anaconda.org/pytorch/magma-cuda115).
```
another nit

    
## 4. Modify scripts to install the new CUDA for Libtorch and Manywheel Docker Linux containers.
There are three types of Docker containers we maintain in order to build Linux binaries: `conda`, `libtorch`, and `manywheel`. They all require installing CUDA and then updating code references in respective build scripts/Dockerfiles.  This step is about libtorch and manywheel containers.

1. Add setup for our Docker `libtorch` and `manywheel`:
    1. For `libtorch`, the code changes are usually copy-paste. For `manywheel`, you should manually verify the versions of the shared libraries with the CUDA you downloaded before.

>     2. Create new repo to host manylinux-cuda images for example : https://hub.docker.com/r/pytorch/manylinux-cuda115 . And give bots write and read acceess to this repo. This step to be removed once following [issue](https://github.com/pytorch/builder/issues/901) is addressed

Reviewer: ```suggestion
    2. Create a new repo to host manylinux-cuda images, for example, https://hub.docker.com/r/pytorch/manylinux-cuda115. Then, give bots write and read access to this repo. This step can be removed once the following [issue](https://github.com/pytorch/builder/issues/901) is addressed.
```

    3. Push the images to Docker Hub. This step should be automated with the help with GitHub Actions in the `pytorch/builder` repo. Make sure to update the `cuda_version` to the version you're adding in respective YAMLs, such as `.github/workflows/build-manywheel-images.yml`, `.github/workflows/build-conda-images.yml`, `.github/workflows/build-libtorch-images.yml`.
    4. Verify that each of the workflows that push the images succeed by selecting and verifying them in the [Actions page](https://github.com/pytorch/builder/actions/workflows/build-libtorch-images.yml) of pytorch/builder. Furthermore, check [https://hub.docker.com/r/pytorch/manylinux-builder/tags](https://hub.docker.com/r/pytorch/manylinux-builder/tags), [https://hub.docker.com/r/pytorch/libtorch-cxx11-builder/tags](https://hub.docker.com/r/pytorch/libtorch-cxx11-builder/tags) to verify that the right tags exist for manylinux and libtorch types of images.

> 

Reviewer: ```suggestion
```
